x-pack/plugins/security/server/session_management/session.ts

```diff
@@ -0,0 +1,431 @@
+      );
+      updateSessionIndex = true;
+    } else if (
+      this.#idleIndexUpdateTimeout !== null &&
+      this.#idleIndexUpdateTimeout <
+        sessionExpirationInfo.idleTimeoutExpiration! - sessionValue.idleTimeoutExpiration!
+    ) {
+      // 3. If idle timeout was updated a while ago.
+      this.#options.logger.debug(
+        'Session idle timeout stored in the index is too old and will be updated.'
+      );
+      updateSessionIndex = true;
+    }
+
+    // First try to store session in the index and only then in the cookie to make sure cookie is
+    // only updated if server side session is created successfully.
+    if (updateSessionIndex) {
+      const sessionIndexValue = await this.#options.sessionIndex.update({
+        ...sessionValue.metadata.index,
+        ...sessionExpirationInfo,
+      });
+
+      // Session may be already invalidated by another concurrent request, in this case we should
+      // clear cookie for the request as well.
+      if (sessionIndexValue === null) {
+        this.#options.logger.warn('Session cannot be extended as it has been invalidated already.');
+        await this.#options.sessionCookie.clear(request);
+        return null;
+      }
+
+      sessionValue.metadata.index = sessionIndexValue;
+    }
+
+    await this.#options.sessionCookie.set(request, {
+      ...sessionCookieValue,
+      ...sessionExpirationInfo,
+    });
+
+    this.#options.logger.debug('Successfully extended existing session.');
+
+    return { ...sessionValue, ...sessionExpirationInfo } as Readonly<SessionValue>;
+  }
+
+  /**
+   * Clears session value for the specified request.
+   * @param request Request instance to clear session value for.
+   */
+  async clear(request: KibanaRequest) {
+    const sessionCookieValue = await this.#options.sessionCookie.get(request);
+    if (!sessionCookieValue) {
+      return null;
+    }
+
+    await Promise.all([
+      this.#options.sessionCookie.clear(request),
+      this.#options.sessionIndex.clear(sessionCookieValue.sid),
+    ]);
+
+    this.#options.logger.debug('Successfully invalidated existing session.');
+  }
+
+  /**
+   * Encrypts session value content and converts to a value stored in the session index.
+   * @param sessionValue Session value.
+   * @param aad Additional authenticated data (AAD) used for encryption.
+   */
+  private async encryptSessionValue(
+    sessionValue: Readonly<Omit<SessionValue, 'metadata'>>,
+    aad: string
+  ) {
+    // Extract values that shouldn't be directly included into session index value.
+    const { username, state, ...sessionIndexValue } = sessionValue;
+
+    try {
+      const encryptedContent = await this.#crypto.encrypt(
+        JSON.stringify({ username, state } as SessionValueContentToEncrypt),
+        aad
+      );
+      return {
+        ...sessionIndexValue,
+        username_hash: username && createHash('sha3-256').update(username).digest('hex'),
```

Comment:
I wonder if we should use a salted hash here. We ship with a number of well-known usernames, and generating an unsalted hash of these won't buy us much. For example, the `elastic` username will always hash to `eb28536c8ead72bf81a0a9226e38fc9bad81f5e07c2081bb801b2a5c8842924e`. It'd be trivial for an attacker to pre-compute these values to see which of these users have active sessions in Kibana.

The problem we have is choosing the salt. We still want a deterministic hash per user so that we have a fighting chance of implementing https://github.com/elastic/kibana/issues/18162, so generating a random salt on the fly is likely out of the question. I hate to leave a dangling comment without a proposed solution, but let me give this some more thought. Open to any ideas you might have too!